File: src-tauri/src/api/repetition_api.rs
use crate::entity::repetition;
use crate::entity::review::Rating;
use crate::model::file_repetitions_count::FileRepetitionCounts;
use crate::service::repetition_service;
use sea_orm::DbConn;
use tauri::State;
use tokio::sync::Mutex;

/// Returns the count of repetitions ready for study, i.e. their due is less
/// than or equal to now.
#[tauri::command]
pub async fn get_study_repetition_counts(
    db_conn: State<'_, Mutex<DbConn>>,
    file_id: i32,
) -> Result<FileRepetitionCounts, String> {
    let db_conn = db_conn.lock().await;
    repetition_service::get_study_repetition_counts(&db_conn, file_id).await
}

#[tauri::command]
pub async fn get_file_repetitions(
    db_conn: State<'_, Mutex<DbConn>>,
    file_id: i32,
) -> Result<Vec<repetition::Model>, String> {
    let db_conn = db_conn.lock().await;
    repetition_service::get_file_repetitions(&db_conn, file_id).await
}

// TODO: move to review_api
#[tauri::command]
pub async fn register_review(
    db_conn: State<'_, Mutex<DbConn>>,
    new_repetition: repetition::Model,
    rating: Rating,
    study_time: i32,
) -> Result<(), String> {
    let db_conn = db_conn.lock().await;
    repetition_service::register_review(&db_conn, new_repetition, rating, study_time).await
}

#[tauri::command]
pub async fn get_repetitions_for_files(
    db_conn: State<'_, Mutex<DbConn>>,
    file_ids: Vec<i32>,
) -> Result<Vec<repetition::Model>, String> {
    let db_conn = db_conn.lock().await;
    repetition_service::get_repetitions_for_files(&db_conn, file_ids).await
}

#[tauri::command]
pub async fn reset_repetitions_for_cell(
    db_conn: State<'_, Mutex<DbConn>>,
    cell_id: i32,
) -> Result<(), String> {
    let db_conn = db_conn.lock().await;
    repetition_service::reset_repetitions_for_cell(&db_conn, cell_id).await
}

File: src/api/repetitionApi.ts
import { invoke } from "@tauri-apps/api/core";
import FileRepetitionCounts from "../type/backend/model/fileRepetitionCounts";
import Repetition from "../type/backend/entity/repetition";
import { Rating } from "../type/backend/entity/rating";

export function getStudyRepetitionCounts(
	fileId: number,
): Promise<FileRepetitionCounts> {
	return invoke("get_study_repetition_counts", {
		fileId,
	});
}

export function registerReview(
	newRepetition: Repetition,
	rating: Rating,
	studyTime: number,
) {
	return invoke("register_review", { newRepetition, rating, studyTime });
}

export function getFileRepetitions(fileId: number): Promise<Repetition[]> {
	return invoke("get_file_repetitions", {
		fileId,
	});
}

export function getRepetitionsForFiles(
	fileIds: number[],
): Promise<Repetition[]> {
	return invoke("get_repetitions_for_files", { fileIds });
}

export function resetRepetitionsForCell(cellId: number) {
	return invoke("reset_repetitions_for_cell", { cellId });
}
